iree/hal/command_buffer.h

```diff
@@ -46,6 +46,27 @@ enum iree_hal_command_buffer_mode_e {
   // TODO(benvanik): IREE_HAL_COMMAND_BUFFER_MODE_REUSABLE = 1u << 1,
   // TODO(benvanik): IREE_HAL_COMMAND_BUFFER_MODE_PRIMARY = 1u << 2,
   // TODO(benvanik): IREE_HAL_COMMAND_BUFFER_MODE_SECONDARY = 1u << 3,
+
+  // Indicates that the command buffer execution is allowed to execute inline
+  // with recording. The exact execution behavior is unspecified by the API and
+  // intentionally unknowable and must always assume to happen entirely
+  // asynchronously and that it will only have completed after waiting on device
+  // idle or the wait semaphores specified in the submission are signaled.
+  //
+  // Local backends can use this to avoid recording when the calling program can
+  // guarantee that it makes no assumptions about execution being deferred until
+  // a submission. The command buffer must still be submitted for scheduling and
+  // must have no wait semaphores specified. This allows the same program code
+  // to execute work both synchronously and asynchronously as remote backends
+  // are allowed to ignore this.
+  //
+  // Remote backends can use this to flush the command buffer more aggressively
+  // to begin early execution and overlap with continued recording.
```

Comment:
yeah I think it'll free up a lot of opportunities without causing massive divergence in outputs. I was really hoping there'd be a way to prevent a giant "command buffer or not" compiler flag and this seems to make that all go away - these inline command buffers on the CPU just become the API for how work is performed, and actually works well to reduce overhead there too. Any compiler improvements for reusing push constant/binding state, descriptor sets, etc will reduce the runtime overhead further than a more naive API that passed a list of buffers/parameters per call!